### ISSUE TITLE
Citation/bibliography support

### ISSUE THREAD

**Issue by Chatatata** [RelaxedJS/ReLaXed#27]
Are you considering to add citation feature, like LaTeX? #8 refers to a ToC support, if it’s gonna be implemented, also adding a bibliography support could be easy as pie.

**Comment by afhoffman** [comment]
Please forgive me if I'm being ignorant...I'm not a JS programmer. It looks like a JS project [exists](https://github.com/pcooksey/bibtex-js) for parsing BibTeX databases already. I'll try to poke at it a little bit, but it looks like this would be a good start for adding citation/bibliography capability.

I think it'd be a good idea if whatever citation capability that's added can directly parse BibTeX databases since most users interested in this feature are probably coming to ReLaXed from LaTeX and are likely already generating their bibliography database files with something like JabRef or Mendeley.

**Comment by DanielRuf** [comment]
https://github.com/larsgw/citation.js/tree/master#starting looks indeed very promising. It uses https://github.com/juris-m/citeproc-js

**Comment by afhoffman** [comment]
I agree. Citation.js looks like it'll do the job. I have some downtime so I'll play with it a little and see what I can come up with.

**Comment by Zulko** [comment]
@afhoffman thanks for the exploration. Some thoughts of this:

- It is great stuff and confirms my thoughts that web techs are progressing faster than latex, even for non-mainstream usecases.
- I am not too optimistic that ReLaXed can fully replace LaTeX any time soon for people wanting to write a full thesis. I would still encourage everyone who has to deal with Bibliography to use LaTeX (that includes me :P)
- The question here is how to make bibliography as easy as LaTeX in Pug. I have an intuition that a lot could be achieved using Pug mixins, as they allow javascript execution. So you would have ``+cite`` and ``+bibliogrpahy`` and ``+bibitem`` mixins and then your code would look like this:

```pug
p The Pug language has been praised in the past (+cite(Zulko2017), +cite(McPugface2015)).

// end of page
+bibliography
```

- I am not sure if bibliography supports needs to be part of the core relaxed project or could be a side project, something that you just import and it gives you bibliography support.

To answer your question @afhoffman some (old) bibliography managers allow you to browse Bibtex files like if they were databases, so having the full abstract was a way to explore and identify the papers you want to cite. This has been made obsolete by Mendeley which connects to the Internet.

**Comment by afhoffman** [comment]
I agree. I think ReLaXed has a ways to go before it could fully replace LaTeX in those cases, but it seems like it'll be much, much easier to use than LaTeX. I don't think anyone will argue that sometimes doing simple things in LaTeX can be too difficult and/or time-consuming. ReLaXed seems like it'll be much more straightforward, even when it's more mature. 

I like the usage example you show there, I think that's exactly what users would want to see when it comes to citation support.

Keeping the bibliography support separate may be a good idea. It'd help you keep the core ReLaXed source code small and users who don't care about bibliography support wouldn't be bothered with it.

**Comment by benperiton** [comment]
>  am not sure if bibliography supports needs to be part of the core relaxed project or could be a side project, something that you just import and it gives you bibliography support.

I think that would be pretty awesome, having some hooks that would allow people to write plugins for certain actions, pre-parse, post-parse, post-render etc

**Comment by Drew-S** [comment]
I was able to accomplish a bibliography using citation-js. It is not a built in solution for relaxed yet, since I got a working system I will clone the repo and see if I cannot get it as a built in system. [Bilbiography setup](https://gist.github.com/Drew-S/afee843d64b3193a975e8c3669962f49)

**Comment by Drew-S** [comment]
We can mark this a closed, now that `citation-js` is implemented by #64.

**Comment by Zulko** [comment]
I noticed that the bibliography system has no footprint (=extra duration) at all even on documents with a HUGE dom, that's pretty remarkable. Thanks @Drew-S ! Next step will be to add proper documentation.
